File: java/src/main/java/com/workshop/architecture/fitness/hexagon/outside/repository/jpa/JpaPlanRepository.java
package com.workshop.architecture.fitness.hexagon.outside.repository.jpa;

import com.workshop.architecture.fitness.hexagon.inside.port.outbound.ForFindingPlans;
import com.workshop.architecture.fitness.hexagon.inside.port.outbound.Plan;
import com.workshop.architecture.fitness.hexagon.inside.port.outbound.PlanNotFoundException;
import com.workshop.architecture.fitness.layered.infrastructure.PlanRepository;
import org.springframework.stereotype.Component;

import java.util.UUID;

@Component
public class JpaPlanRepository implements ForFindingPlans {
    private final PlanRepository planRepository;

    public JpaPlanRepository(PlanRepository planRepository) {
        this.planRepository = planRepository;
    }
    
    public Plan findPlanById(UUID planId) throws PlanNotFoundException {
        var planEntity = planRepository.findById(planId)
                .orElseThrow(() -> new PlanNotFoundException(
                        "Plan %s was not found".formatted(planId.toString())
                ));
        return new Plan(planId, planEntity.getPrice(), planEntity.getDurationInMonths(), planEntity.getTitle());
    }
}

File: python/src/workshop_api/fitness/hexagon/outside/driven/repository/jpa/jpa_plan_repository.py
from uuid import UUID

from workshop_api.fitness.hexagon.inside.port.outbound.for_finding_plans import (
    ForFindingPlans,
)
from workshop_api.fitness.hexagon.inside.port.outbound.plan import Plan
from workshop_api.fitness.hexagon.inside.port.outbound.plan_not_found_exception import (
    PlanNotFoundException,
)
from workshop_api.fitness.layered.infrastructure.plan_repository import PlanRepository


class JpaPlanRepository(ForFindingPlans):
    def __init__(self, plan_repository: PlanRepository) -> None:
        self.plan_repository = plan_repository

    def find_plan_by_id(self, plan_id: UUID) -> Plan:
        plan_entity = self.plan_repository.find_by_id(plan_id)
        if plan_entity is None:
            raise PlanNotFoundException(f"Plan {plan_id} was not found")

        return Plan(
            id=plan_id,
            price=plan_entity.price,
            duration_in_months=plan_entity.duration_in_months,
            title=plan_entity.title,
        )
